Python/compile.c

```diff
@@ -3843,7 +3843,8 @@ compiler_nameop(struct compiler *c, identifier name, expr_context_ty ctx)
             optype = OP_FAST;
         break;
     case GLOBAL_IMPLICIT:
-        if (c->u->u_ste->ste_type == FunctionBlock)
+        if (c->u->u_ste->ste_type == FunctionBlock ||
```

Comment:
Not confident this is right because I don't yet know what GLOBAL_IMPLICIT does.

Reply:
I'm not super-duper confident, but the sense I get is that GLOBAL_IMPLICIT means "this name wasn't declared global using the global keyword, but contextually we know it has to be a global".

Reply:
or maybe it's better described as "we've ruled out it all the other possibilities, the only option left is that it's GLOBAL".

Reply:
I tackled it from another perspective and things seem clearer.

`GLOBAL_IMPLICIT` is set in two places, but it's only ever used once, in the spot you've modified above.  The code in question decides what type of access to use for the name:

* `FAST`,
* `DEREF`,
* `NAME`, or
* `GLOBAL`.

In order to join that access type with the verb:

* `Load`,
* `Store`,
* or `Del`.

To construct the final opcode, e.g.:

* `LOAD_FAST`,
*`STORE_NAME`,
* et cetera.

If the name is of type `GLOBAL_EXPLICIT`, the access type is always `GLOBAL`.  If it's `GLOBAL_IMPLICIT`, if the scope/block it's in is a `FunctionBlock`, the access type is `GLOBAL`, otherwise (in `ClassBlock` and `ModuleBlock`) it defaults to `NAME`.

Given that, I *think* I know what the right thing to do here is.  Looking up a variable in an `AnnotationBlock` should behave as if the lookup happened in the enclosing scope.  Which means that if we're an `AnnotationBlock` inside a `FunctionBlock`, `GLOBAL_IMPLICIT` should resolve to `GLOBAL` access.  If we're an `AnnotationBlock` inside a `ClassBlock` or `ModuleBlock`, it should resolve to `NAME` access.

p.s. the code that sets `GLOBAL_IMPLICIT` at the end of `analyze_name()` in `symtable.c` is needlessly abstruse.  Maybe we should change it to

    if (ste->ste_nested && !(global && PySet_Contains(global, name))) {
        ste->ste_free = 1;
    }
    SET_SCOPE(scopes, name, GLOBAL_IMPLICIT);
    return 1;

Reply:
So that argues that the change you commented on above isn't quite right.  Instead, we need a new field in `_symtable_entry`, let's call it `ste_parent`, which for an `AssertionBlock` points to the symbol table of the enclosing scope.  Then we revert your change above, and instead insert four new lines under `case GLOBAL_IMPLICIT:` as follows:

    case GLOBAL_IMPLICIT:
	_Py_block_ty ste_type = c->u->u_ste->ste_type;
	if (ste_type == AnnotationBlock) {
	    assert(c->u->u_ste->ste_parent);
	    ste_type = c->u->u_ste->ste_parent->ste_type;
	}
	if (ste_type == FunctionBlock)

Reply:
Oh, dear, that's not quite right either.  If we're an annotation on a method of a class, then the `AnnotationBlock` is enclosed by a `ClassBlock`.  But that would mean defaulting to `NAME` access, which isn't what we want, because the namespace of the class block is gone by then.  So let me amend what I said: an `AnnotationBlock` needs to have *visibility* on variables as if the names are being resolved in the enclosing scope/block.  But it obviously should use whatever mechanism is appropriate to achieve that.  In the scenario I just described, an annotation to a parameter of a method defined in a class, it needs to be a closure and use `DEREF`.  So, great!, now your code is correct :D

Reply:
Though now I worry we're wading into uncharted territory.  It's not possible for functions defined inside classes to see the names from the class scope.  And functions that refer to globals just look up the name in globals.  So I bet in Python there is currently literally no way to create a `DEREF` that refers to a name using `NAME` access.  I genuinely still don't know how cell variables work, so I don't know if this is something where it already works for free, or whether we're gonna have to go write it.  Going AFK for a couple hours, will stare at it after that.